httpclient/httpclient.go

```diff
@@ -6,23 +6,118 @@ Released under MIT license.
 
 package httpclient
 
-import "net/http"
-
-// CloneHTTPRequest creates a shallow copy of the request along with a deep copy of the Headers.
-func CloneHTTPRequest(req *http.Request) *http.Request {
-	r := new(http.Request)
-	*r = *req
-	r.Header = CloneHTTPHeader(req.Header)
-	return r
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/acronis/go-appkit/log"
+)
+
+// New wraps delegate transports with logging, rate limiting, retryable, request id
+// and returns an error if any occurs.
+func New(cfg *Config) (*http.Client, error) {
+	return NewWithOpts(cfg, Opts{})
+}
+
+// Must wraps delegate transports with logging, rate limiting, retryable, request id
+// and panics if any error occurs.
+func Must(cfg *Config) *http.Client {
+	client, err := New(cfg)
+	if err != nil {
+		panic(err)
+	}
+
+	return client
+}
+
+// Opts provides options for NewWithOpts and MustWithOpts functions.
+type Opts struct {
+	// UserAgent is a user agent string.
+	UserAgent string
+
+	// ClientType represents a type of client, it's a service component reference. e.g. 'auth-service'
+	ClientType string
+
+	// Delegate is the next RoundTripper in the chain.
+	Delegate http.RoundTripper
+
+	// LoggerProvider is a function that provides a context-specific logger.
+	LoggerProvider func(ctx context.Context) log.FieldLogger
+
+	// RequestIDProvider is a function that provides a request ID.
+	RequestIDProvider func(ctx context.Context) string
+
+	// Collector is a metrics collector.
+	Collector MetricsCollector
```

Comment:
I suggest renaming `Collector` field to MetricsCollector`. Otherwise, it’s not immediately clear from the field’s name what this collector is for.